sklearn/pipeline.py

```diff
@@ -253,33 +254,44 @@ def set_metadata_request(self, props):
                            "level and not at the pipeline level.")
 
     def _check_fit_params(self, **fit_params):
-        required_props = self.get_metadata_request().fit
-        print("fit params steps:", list(fit_params.keys()))
-        _validate_required_props(required_props, fit_params)
-
         fit_params_steps = {name: {} for name, step in self.steps
                             if step is not None}
-        """
-        for pname, pval in fit_params.items():
-            if '__' not in pname:
-                raise ValueError(
-                    "Pipeline.fit does not accept the {} parameter. "
-                    "You can pass parameters to specific steps of your "
-                    "pipeline using the stepname__parameter format, e.g. "
-                    "`Pipeline.fit(X, y, logisticregression__sample_weight"
-                    "=sample_weight)`.".format(pname))
-            step, param = pname.split('__', 1)
-            fit_params_steps[step][param] = pval
-        """
+        # Remove old_behavior in 0.26
+        old_behavior = np.any(['__' in pname for pname in fit_params.keys()])
+        if old_behavior:
+            warnings.warn("It seems fit_params are using the deprecated "
+                          "way to route parameters using '__'. This behavior "
+                          "is deprecated in 0.24 and won't be accepted in "
+                          "0.26. Please use `set_request_metadata` to route "
+                          "parameters.", FutureWarning)
+            for pname, pval in fit_params.items():
+                if '__' not in pname:
+                    raise ValueError(
+                        "Pipeline.fit does not accept the {} parameter. "
+                        "You can pass parameters to specific steps of your "
+                        "pipeline using the stepname__parameter format, e.g. "
+                        "`Pipeline.fit(X, y, logisticregression__sample_weight"
+                        "=sample_weight)`.".format(pname))
+                step, param = pname.split('__', 1)
+                fit_params_steps[step][param] = pval
+            return fit_params_steps
+
+        required_props = self.get_metadata_request().fit
+        # print("fit params steps:", list(fit_params.keys()))
+        _validate_required_props(required_props, fit_params)
         for _, name, transformer in self._iter(filter_passthrough=True):
             if transformer is None:
                 continue
-            transformer_fit_props = transformer.get_metadata_request().fit
+            try:
+                transformer_fit_props = transformer.get_metadata_request().fit
```

Comment:
Do we want to require attribute access to get fit?

Reply:
don't understand, could you please elaborate?

Reply:
I mean this assumes that the metadata request has an attribute called 'fit'. Should we assume that?

Reply:
`get_metadata_request()` will always return a `fit` attribute even if it's empty with the current implementation.